### PR TITLE
Add timeout for image pulling

### LINKED ISSUE #26300: Set timeout for docker image pull calls
We recently removed the timeout for image pulls because the default is too low (i.e. 2 min), and it's hard to gauge what time will work for the users.

However, it's important that we set "a certain" timeout for such requests to resolve the cases where the request hangs indefinitely.
Off the top of my head, there are two options:

 1. Add the timeout back. Set the default to something longer (e.g., 10 minutes) and make it configurable through kubelet flags.
 2. Detect whether the pull request is making progress by leveraging the pull progress reporter (#26145) and cancel the request if the progress has stalled over a threshold.

/cc @kubernetes/sig-node

### PR DESCRIPTION
**Pull request by Random-Liu**
Fix #26300.

With this PR, if image pulling makes no progress for _1 minute_, the operation will be cancelled. Docker reports progress for every 512kB block (See [here](https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/pkg/progress/progressreader.go#L32)), _512kB/min_ means the throughput is _<= 8.5kB/s_, which should be kind of abnormal?

It's a little hard to write unit test for this, so I just manually tested it. If I set the `defaultImagePullingStuckTimeout` to 0s, and `defaultImagePullingProgressReportInterval` to 1s, image pulling will be cancelled.

```
E0601 18:48:29.026003   46185 kube_docker_client.go:274] Cancel pulling image "nginx:latest" because of no progress for 0, latest progress: "89732b811e7f: Pulling fs layer "
E0601 18:48:29.026308   46185 manager.go:2110] container start failed: ErrImagePull: net/http: request canceled
```

/cc @kubernetes/sig-node 
[Analytics]()

### CODE REVIEW

File: pkg/kubelet/dockertools/kube_docker_client.go [hunk 1]
```diff
@@ -270,14 +297,14 @@ func (d *kubeDockerClient) PullImage(image string, auth dockertypes.AuthConfig,
 		return err
 	}
 	opts.RegistryAuth = base64Auth
-	// Don't set timeout for the context because image pulling can be
-	// take an arbitrarily long time.
-	resp, err := d.client.ImagePull(context.Background(), image, opts)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/kube_docker_client.go]: question: calling `cancel` multiple times should not have any negative impact, right?

File: pkg/kubelet/dockertools/kube_docker_client.go [hunk 2]
```diff
@@ -203,55 +209,76 @@ type progress struct {
...
 
-func (p *progress) get() string {
+func (p *progress) get() (string, time.Time) {
 	p.RLock()
 	defer p.RUnlock()
 	if p.message == nil {
-		return "No progress"
+		return "No progress", p.timestamp
 	}
+	// The following code is migrated from JSONMessage.Display
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/kube_docker_client.go]: s/migrated/copied or s/migrated from/based on

File: pkg/kubelet/dockertools/kube_docker_client.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/kube_docker_client.go]: Yeah.
```
A CancelFunc tells an operation to abandon its work. A CancelFunc does not wait for the work to stop. After the first call, subsequent calls to a CancelFunc do nothing.
```

File: pkg/kubelet/dockertools/kube_docker_client.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/kube_docker_client.go]: ACK.

**Comment by ashahba** [comment]
Hi,
Question about this PR: What if the image pull is temporarily slow, say for 1 minute, this causes the image pull to fail, although the issue might be only temporary?
This could cause some unpleasant failures for customers which I don't think the PR meant to introduce.

**Comment by yujuhong** [comment]
> Question about this PR: What if the image pull is temporarily slow, say for 1 minute, this causes the image pull to fail, although the issue might be only temporary?
> This could cause some unpleasant early failures for customers and I don't think the PR meant to introduce that issue.

It's not frequent that image pulling would be stuck for over 1 minute with no progress at all, and still recover in the end. Without this PR, kubelet may end up having many requests hanging forever.
If the issue is temporary, kubelet will retry and succeed next time. If you have the need for a higher timeout, you can customize it through a flag.

> Also is this PR unintentionally forcing people to use GCR as their Docker registry?

I have no idea how you drew the conclusion. Care to elaborate?

**Pull request merged**